### ISSUE TITLE
gcp_cloud_storage sink token refresh failing

### ISSUE THREAD

**Issue by srinidhis94** [vectordotdev/vector#13058]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Hi,
    I am running vector as a daemon-set. We are running a big cluster with approximately 700+ nodes running. When we deploy vector, we initially see ratelimiting error. Looks like it is from k8s server. Find the log below. We get this logs approximately for 3-4 minutes before it resolves and starts working. Also there's no back off  here as we see continuous stream of below error.
    
  `2022-06-09T07:28:18.523823Z  WARN vector::kubernetes::reflector: Watcher Stream received an error. Retrying. error=InitialListFailed(Api(ErrorResponse { status: "429 Too Many Requests", message: "\"Too many requests, please try again later.\\n\"", reason: "Failed to parse error data", code: 429 }))`
  
  Once the above error resolves, vector works without any issues for 55 minutes and uploads logs to gcs bucket correctly. Then we start seeing request unauthorized error for gcp_cloud_storage requests and never recovers. Find the error log below
  
  `2022-06-09T08:25:51.906165Z ERROR sink{component_kind="sink" component_id=cloud_storage component_type=gcp_cloud_storage component_name=cloud_storage}:request{request_id=803}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 401 Unauthorized"`
  
  Currently I am suspecting if this issue is a side effect of initial rate limit error. I am deleting pods slowly so that new pods come up and their requests don't get ratelimited. Will comment my observation in sometime.

### Configuration

```text
We're deploying it using latest helm chart available. Have added vector image version.

cloud_storage_analytics:
  type: gcp_cloud_storage
  inputs: [ test_logs ]
  compression: gzip
  batch:
    max_bytes: 5000000
    timeout_secs: 60
  encoding:
    codec: ndjson
  bucket: test-bucket
  key_prefix: 'analytics/dam/default/{{`{{ component_name }}`}}/year=%Y/month=%m/day=%d/hour=%H/%M_'
```

### Version

0.22.0-distroless-libc

### Debug Output

```text
2022-06-09T07:28:18.523823Z  WARN vector::kubernetes::reflector: Watcher Stream received an error. Retrying. error=InitialListFailed(Api(ErrorResponse { status: "429 Too Many Requests", message: "\"Too many requests, please try again later.\\n\"", reason: "Failed to parse error data", code: 429 }))


2022-06-09T08:25:51.906165Z ERROR sink{component_kind="sink" component_id=cloud_storage component_type=gcp_cloud_storage component_name=cloud_storage}:request{request_id=803}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 401 Unauthorized"
```

### Example Data

2022-06-09T07:28:18.523823Z  WARN vector::kubernetes::reflector: Watcher Stream received an error. Retrying. error=InitialListFailed(Api(ErrorResponse { status: "429 Too Many Requests", message: "\"Too many requests, please try again later.\\n\"", reason: "Failed to parse error data", code: 429 }))

2022-06-09T08:25:51.906165Z ERROR sink{component_kind="sink" component_id=cloud_storage component_type=gcp_cloud_storage component_name=cloud_storage}:request{request_id=803}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 401 Unauthorized"

**Comment by srinidhis94** [comment]
Hi,
    I tried deleting each pod in vector daemonset one by one with a sleep time of 2 seconds. This was done to ensure not ratelimits are hit on k8s APIs used by vector. After this exercise, there's definitely no issues wrt authorisation.  This has something to do with failure in error handling during initialisation of vector. Can you guys have a look at this ? I think you can replicate this by starting 700+ vector daemonsets at the same time which results in failure to initialise due to k8s ratelimit errors.

**Comment by spencergilbert** [comment]
Hi @srinidhis94 - it seems unlikely that the WARNs from the `kubernetes_logs` source is affecting the sink, are you able to reproduce this outside of kubernetes?

**Comment by srinidhis94** [comment]
Hi @spencergilbert , Currently we run only in k8s environment. So I haven't tried any other sources. But within k8s environment, I can reproduce and I can provide more logs if required.

**Comment by spencergilbert** [comment]
Can you run with `VECTOR_LOG=vector::gcp=debug,gouth=debug,goauth=debug` - this should log any events related to the GCP tokens

**Comment by srinidhis94** [comment]
Sure. Will try in our test environment and update you.

**Comment by srinidhis94** [comment]
Hi @spencergilbert , while trying to reproduce this issue, we faced another problem. But that is pretty straight forward.
Vector is currently using /pods API for identifying pods from resources.
https://vector.dev/docs/reference/configuration/sources/kubernetes_logs/#kubernetes-api-access-control

K8S has 2 APIs to get cluster pod details.
Endpoint API :
kubectl get   --raw '/api/v1/namespaces/default/pods?watch=1'

Endpoint slices :
kubectl get --raw /apis/discovery.k8s.io/v1/watch/namespaces/default/endpointslices

The default endpoints API which is currently being used chokes the network at scale and hence fails to list pods as well as fetch logs. A better alternate is endpoint slices. More document can be found below.
https://kubernetes.io/blog/2020/09/02/scaling-kubernetes-networking-with-endpointslices/

**Comment by spencergilbert** [comment]
👍 would you mind opening an issue to track this?

**Comment by fraserdarwent** [comment]
@spencergilbert @bruceg related to https://github.com/vectordotdev/vector/issues/10828 ?

**Comment by spencergilbert** [comment]
Ah, most likely - @srinidhis94 have you seen this after upgrading to the most recent version?

**Comment by punkerpunker** [comment]
@spencergilbert I can confirm that the error persists in version `0.24`. I am using official helm charts to spin it up on kubernetes. Same behavior as described - after ~1 hour token doesn't seem to be refreshing, though requests are failed with 401 error

```
2022-10-10T10:44:37.312145Z DEBUG sink{component_kind="sink" component_id=gcs_predictor component_type=gcp_cloud_storage component_name=gcs_predictor}:request{request_id=3}:http: vector::internal_events::http_client: HTTP response. status=401 Unauthorized version=HTTP/1.1 headers={"x-guploader-uploadid": "ADPycdtfwMU3n0G-S8nFvs6HA_L4RII_em74m0ZLl68_oong_BEI1sXYgck9NHYn_caiXRLPeUhqRAo176wjZFDbUcV3-rAejsZ8", "content-type": "application/xml; charset=UTF-8", "content-length": "131", "www-authenticate": "Bearer realm=\"https://accounts.google.com/\", error=\"invalid_token\"", "date": "Mon, 10 Oct 2022 10:44:37 GMT", "server": "UploadServer"} body=[131 bytes]
2022-10-10T10:44:37.312193Z ERROR sink{component_kind="sink" component_id=gcs_predictor component_type=gcp_cloud_storage component_name=gcs_predictor}:request{request_id=3}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 401 Unauthorized"
```

**Comment by punkerpunker** [comment]
I took a proper look at what was causing the above issue and seems like the disabled healthcheck is the reason.

According to what I see here: 
https://github.com/vectordotdev/vector/blob/a4782f83b0fa3ac8d6ca1f56a4990687e338ed06/src/sinks/gcs_common/config.rs#L128-L144

Token regenerator is being spawned only after the health check response is received (good to mention - it is spawned regardless of the healtcheck status). 

So for me, **enabling a healthcheck** was the fix. Quite obscure though, are we expecting this behaviour or shall we add a warning in documentation or open up a PR with the fix? 

cc @spencergilbert @srinidhis94

**Comment by spencergilbert** [comment]
@bruceg I believe you implemented this originally, I'm guessing this wasn't intended behavior (or something I'm missing as a requirement). What do you think?

**Comment by z2665** [comment]
I also encountered a situation where I received an error after sinking to gcs and running for a while

```
ERROR sink{component_kind="sink" component_id=gcs_logs component_type=gcp_cloud_storage component_name=gcs_logs}:request{request_id=1474}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 401 Unauthorized"
```

But I don't actively disable healthcheck in sink, I see that it is enabled by default in the config file and I have the only log that healthcheck is enabled

```
INFO vector::topology::running: Running healthchecks.
```

**Comment by jszwedko** [comment]
Closed by https://github.com/vectordotdev/vector/pull/15827 . Will be included in 0.28.0 release.

**Comment by jszwedko** [comment]
Closed by https://github.com/vectordotdev/vector/pull/15827
